### DISCUSSION TITLE
False positive workings with: unknown field in rust-toolchain.toml: toolchain

### DISCUSSION THREAD

**Discussion by ssbarnea** [jdx/mise#4005] [Troubleshooting and bug reports]
It seems that there is some wrong logic related to processing `rust-toolchain.toml` which can cause some false positive warnings. Apparently, `mise` is trying to load these as `mise.toml` files and they do not have the same format.

The most hilarious part is that the warning does not appear when I enter a directory that have this file but it appears when I exit that directory and switch to another one.

```shell
❯ cd uv
~/code/os/uv

❯ cd tox-uv
~/code/os/tox-uv
mise WARN  unknown field in ~/code/os/uv/rust-toolchain.toml: toolchain

❯ cat ~/code/os/uv/rust-toolchain.toml
[toolchain]
```

Look at the path reported, warning appearing in another unrelated folder. I should mention that I do not have any rust plugins installed.

```shell
❯ mise plugins
uv
yarn
```

**Comment by roele** [comment]
Can you post your `mise doctor` output to give some more context on your setup?

**Comment by ssbarnea** [comment]
```
~/code/os/uv
❯ mise doctor
version: 2024.12.20 macos-arm64 (2024-12-25)
activated: yes
shims_on_path: no

build_info:
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS
  Built: Wed, 25 Dec 2024 20:45:03 +0000
  Rust Version: rustc 1.83.0 (90b35a623 2024-11-26) (Homebrew)
  Profile: release

shell:
  /bin/zsh
  zsh 5.9 (arm64-apple-darwin24.0)

dirs:
  cache: ~/.cache/mise
  config: ~/.config/mise
  data: ~/.config/mise
  shims: ~/.config/mise/shims
  state: ~/.local/state/mise

config_files:
  ~/.config/mise/config.toml
  ~/code/os/uv/rust-toolchain.toml

ignored_config_files: (none)

backends:
  aqua
  asdf
  cargo
  core
  dotnet
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:
  uv    https://github.com/asdf-community/asdf-uv.git#8155fc5
  yarn  https://github.com/mise-plugins/asdf-yarn.git#74ea3b9

toolset:
  asdf:yarn@4.5.3
  core:java@23.0.1
  core:node@22.12.0
  core:python@3.10.15
  core:python@3.11.11
  core:python@3.12.7
  core:python@3.13.1
  core:python@3.9.21
  core:rust@1.83.0      (missing)
  npm:cspell-cli@8.17.0
  npm:prettier@3.4.2
  ubi:jdx/usage@1.7.4

path:
  ~/.venv/bin
  ~/.config/mise/installs/java/23.0.1/bin
  ~/.config/mise/installs/node/22.12.0/bin
  ~/.config/mise/installs/npm-prettier/3.4.2/bin
  ~/.config/mise/installs/npm-cspell-cli/8.17.0/bin
  ~/.config/mise/installs/python/3.13.1/bin
  ~/.config/mise/installs/python/3.12.7/bin
  ~/.config/mise/installs/python/3.11.11/bin
  ~/.config/mise/installs/python/3.10.15/bin
  ~/.config/mise/installs/python/3.9.21/bin
  ~/.config/mise/installs/usage/1.7.4/bin
  ~/.config/mise/installs/yarn/4.5.3/bin
  /opt/homebrew/opt/openjdk/bin
  ~/bin
  ~/.local/bin
  /usr/local/sbin
  /usr/local/bin
  /Library/Frameworks/Python.framework/Versions/3.12/bin
  /opt/homebrew/bin
  /opt/homebrew/sbin
  /System/Cryptexes/App/usr/bin
  /usr/bin
  /bin
  /usr/sbin
  /sbin
  /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin
  /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin
  /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
  /opt/X11/bin
  /Library/Apple/usr/bin
  /usr/local/MacGPG2/bin
  /usr/local/share/dotnet
  ~/.dotnet/tools
  /Library/Frameworks/Mono.framework/Versions/Current/Commands
  /usr/local/munki
  /Applications/iTerm.app/Contents/Resources/utilities

env_vars:
  MISE_SHELL=zsh

settings:
  experimental         true ~/.config/mise/config.toml
  python.uv_venv_auto  true ~/.config/mise/config.toml
mise WARN  mise version 2024.12.21 available

1 problem found:

1. new mise version 2024.12.21 available, currently on 2024.12.20

zsh (ssbarnea) 3.13 main ↓4  -1  1 (ERROR)~/code/os/uv
❯ cd tox-uv
~/code/os/tox-uv
mise WARN  unknown field in ~/code/os/uv/rust-toolchain.toml: toolchain
```

**Comment by roele** [comment]
This seems like a very odd issue. I am so far not able to reproduce this behaviour. 

Generally `rust-toolchain.toml` is used as idiomatic version file (like .nvmrc etc.), so the file being listed as config file is correct. The file is not meant to be parsed as `mise.toml` as the core Rust plugin provides an implementation to parse the file to return the requested Rust version. I can also not fine any code that attempts to do this obviously (maybe non-obviously using a trait default implementation or something). Your toolchain reports Rust as missing, so your global config file references a `rust@1.83.0`? The `rust-toolchain.toml` does not seem to contain a version as per your `cat` command.

Can you maybe also share your global `mise.toml` and `rust-toolchain.toml` file?

**Comment by fiadliel** [comment]
@jdx @roele
I just came across this. It turns out that the trigger is setting `experimental = true`, it adds this line to `mise config ls`:
```
~/[...]/rust-toolchain.toml  rust
```

This is caused by the `idiomatic_filenames` implementation of `Backend` for `plugins::core::rust::RustPlugin`, which adds `rust-toolchain.toml` to the list of "idiomatic filenames".

The bit of the trace around the error:
```
TRACE  1 [src/config/config_file/mise_toml.rs:113] parsing: ~/[...]/subdir/mise.toml
TRACE  1 [src/cli/args/tool_arg.rs:60] parsing ToolVersionType from: latest
TRACE  1 [src/cli/args/tool_arg.rs:60] parsing ToolVersionType from: 24
TRACE  1 [src/file.rs:176] cat ~/[...]/subdir/rust-toolchain.toml
TRACE  1 [src/config/config_file/mise_toml.rs:113] parsing: ~/[...]/subdir/rust-toolchain.toml
WARN   1 [src/config/config_file/mise_toml.rs:116] unknown field in ~/[...]/subdir/rust-toolchain.toml: toolchain
TRACE  1 [src/cli/mod.rs:342] run hook-env done
TRACE  1 [src/main.rs:88] main done
```

This is when changing from `subdir` to its parent directory.

I think this code may be responsible in `config::config_file`:
```rust
fn detect_config_file_type(path: &Path) -> Option<ConfigFileType> {
    match path
        .file_name()
        .and_then(|f| f.to_str())
        .unwrap_or("mise.toml")
    {
        f if f.ends_with(".toml") => Some(ConfigFileType::MiseToml),
```

**Reply by roele** [reply]
Thanks for the detailed analysis @fiadliel. A fix is underway with https://github.com/jdx/mise/pull/4321.

**Reply by fiadliel** [reply]
Thanks! 🥳

**Comment by Marukome0743** [comment]
Fixed in v2025.2.2 by #4321 (ba43871), merged 2025-02-08 — confirming here since the thread ended while it was still "underway".

@fiadliel's diagnosis was right on both counts. The fix was to move the `idiomatic_filenames` arm of `detect_config_file_type` above the `f if f.ends_with(".toml")` catch-all, so `rust-toolchain.toml` classifies as `IdiomaticVersion` instead of falling through to `MiseToml` and being handed to the mise.toml parser. That ordering still holds today — the check now lives in `path_is_idiomatic()` in `src/config/config_file/mod.rs`.

Worth adding, since it landed after this thread went quiet: #6501 (v2025.10.0) made idiomatic version files opt-in. `idiomatic_version_file_enable_tools` is empty by default now, so `rust-toolchain.toml` is not treated as a config file at all unless you enable it. The `experimental = true` trigger identified here no longer reaches that code path on default settings.

Verified on 2026.7.15 with `experimental = true` set globally.

Default settings:

```
$ cat ~/uv/rust-toolchain.toml
[toolchain]
channel = "1.83.0"

$ cd ~/uv && mise cfg ls
~/.config/mise/config.toml  (none)

$ cd ~/tox-uv && mise hook-env -s bash    # leaving the directory: the original trigger
(no warning)
```

With the tool opted in, the file is now read by the Rust plugin rather than parsed as mise.toml:

```
$ export MISE_IDIOMATIC_VERSION_FILE_ENABLE_TOOLS=rust

$ cd ~/uv && mise cfg ls
~/.config/mise/config.toml  (none)
~/uv/rust-toolchain.toml    rust

$ mise ls rust
rust  1.83.0 (missing)  ~/uv/rust-toolchain.toml  1.83.0

$ cd ~/tox-uv && mise hook-env -s bash
(no warning)
```

And with the literal file from your report, which has no `channel` key:

```
$ cat ~/uv/rust-toolchain.toml
[toolchain]

$ cd ~/uv && mise cfg ls
~/.config/mise/config.toml  (none)
~/uv/rust-toolchain.toml    (none)

$ cd ~/tox-uv && mise hook-env -s bash
(no warning)
```

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.220.*
